dss/error.py

```diff
@@ -42,27 +44,34 @@ def dss_exception_handler(e: DSSException) -> FlaskResponse:
             'stacktrace': traceback.format_exc(),
         })
 
-
 def dss_handler(func):
     @functools.wraps(func)
     def wrapper(*args, **kwargs):
-        try:
-            return func(*args, **kwargs)
-        except werkzeug.exceptions.HTTPException as ex:
-            status = ex.code
-            code = ex.name
-            title = str(ex)
-            stacktrace = traceback.format_exc()
-        except DSSException as ex:
-            status = ex.status
-            code = ex.code
-            title = ex.message
-            stacktrace = traceback.format_exc()
-        except Exception as ex:
-            status = requests.codes.server_error
-            code = "unhandled_exception"
-            title = str(ex)
-            stacktrace = traceback.format_exc()
+        if (os.environ.get('DSS_READ_ONLY_MODE') is None or
+                "GET" == request.method or
+                ("POST" == request.method and "search" in request.path)):
+            try:
+                return func(*args, **kwargs)
+            except werkzeug.exceptions.HTTPException as ex:
+                status = ex.code
+                code = ex.name
+                title = str(ex)
+                stacktrace = traceback.format_exc()
+            except DSSException as ex:
+                status = ex.status
+                code = ex.code
+                title = ex.message
+                stacktrace = traceback.format_exc()
+            except Exception as ex:
+                status = requests.codes.server_error
+                code = "unhandled_exception"
+                title = str(ex)
+                stacktrace = traceback.format_exc()
+        else:
+            status = requests.codes.not_allowed
+            code = "read_only"
+            title = "The DSS is currently read-only"
```

Comment:
Consider changing this to a 5xx series error code, and adding a Retry-After header.